### DISCUSSION TITLE
Will upgrading NDK upgrade the zlib version?

### DISCUSSION THREAD

**Discussion by TedaLIEz** [android/ndk#1968] [Q&A] [answered]
Suppose I have a CMakeLists.txt to build an Android library with native C++ code:
```
find_library(
        zlib
        z
)
```

The zlib library should be a prebuilt library provided by the Android system, I'm wondering if we can upgrade the zlib version (for example from 1.2.11 to 1.2.13) via upgrading our NDK tool version?

This confuses me a lot because it is said that the zlib library is prebuilt and located on the Android device. Besides, I checked my library artifact, I didn't find the zlib.so included.

**Comment by enh-google** [comment, marked as answer]
"it's complicated". (and the result of a historical accident --- if we were creating the NDK today, we just wouldn't include libz.)

* the _headers_ get updated every NDK release.

* the libz.a _static_ library gets updated every NDK release.

* but the libz.so _shared_ library comes from the device, so "you get what you get".

you're right that https://developer.android.com/ndk/guides/stable_apis#zlib_compression could be more helpful. i'll fix that today...

**Reply by enh-google** [reply]
(documentation updated now.)

**Reply by TedaLIEz** [reply]
@enh-google When I investigate the NDK folder downloaded via SDK manager in my build machine, I can find some libz.so files, for example, `ndk\26.1.10909125\toolchains\llvm\prebuilt\windows-x86_64\sysroot\usr\lib\arm-linux-androideabi\34\libz.so`

I'm wondering when I use find_library in CMakeLists.txt, does cmake link to the libz.so inside of the NDK? If so, we should align to the version of zlib.so instead of the so file in the device.

**Reply by TedaLIEz** [reply]
And I searched in my apk file, I didn't find the libz.so in my apk file neither.

**Reply by enh-google** [reply]
> When I investigate the NDK folder downloaded via SDK manager in my build machine, I can find some libz.so files

yes, but those .so files aren't "real" --- they're just stubs because the linker on the machine you're building on needs _something_ to check whether functions/variables exist or not. at _runtime_ you'll be using the ones on the device. this is true for all the NDK .so files.

> I'm wondering when I use find_library in CMakeLists.txt, does cmake link to the libz.so inside of the NDK? If so, we should align to the version of zlib.so instead of the so file in the device.

i think you're confusing "linking" and "loading" (also known as "dynamic linking", confusingly). at *link* time (done by lld at build time), yes, the NDK .a or .so file is used. in the case of the .a file, that's what gets copied directly into your resulting output .so file. in the case of the .so file, your resulting output .so file just has a `DT_NEEDED` entry saying "at runtime, i need libz.so". then at _run_ time, the dynamic linker (aka loader aka /system/bin/linker64) opens your .so file, sees that `DT_NEEDED`, looks for libz.so, finds the one on the system, and points all calls to libz functions to that.

--------------

i think it might be more useful for both of us if you try to rephrase your question as a regular bug report:
1. what did you do?
2. what did you see?
3. what did you expect to see?

it's not clear to me yet what your problem is/what problem you're trying to solve. (that is: "why do you care about the answer to this question? what's broken?" :-) )
